pkg/scheduler/framework/framework.go

```diff
@@ -296,17 +297,186 @@ func (f *framework) markAsUnscheduledFor(ctx context.Context, bindings []*fleetv
+	return scored, filtered, nil
+}
+
+// runPreFilterPlugins runs all pre filter plugins sequentially.
+func (f *framework) runPreFilterPlugins(ctx context.Context, state *CycleState, policy *fleetv1beta1.ClusterSchedulingPolicySnapshot) *Status {
+	for _, pl := range f.profile.preFilterPlugins {
+		status := pl.PreFilter(ctx, state, policy)
+		switch {
+		case status.IsSuccess(): // Do nothing.
+		case status.IsInteralError():
+			return status
+		case status.IsSkip():
+			state.skippedFilterPlugins.Insert(pl.Name())
+		default:
+			// Any status that is not Success, InternalError, or Skip is considered an error.
+			return FromError(fmt.Errorf("prefilter plugin returned an unknown status %s", status), pl.Name())
+		}
+	}
+
+	return nil
+}
+
+// runFilterPluginsFor runs filter plugins for a single cluster.
+func (f *framework) runFilterPluginsFor(ctx context.Context, state *CycleState, policy *fleetv1beta1.ClusterSchedulingPolicySnapshot, cluster *fleetv1beta1.MemberCluster) *Status {
+	for _, pl := range f.profile.filterPlugins {
+		// Skip the plugin if it is not needed.
+		if state.skippedFilterPlugins.Has(pl.Name()) {
+			continue
+		}
+		status := pl.Filter(ctx, state, policy, cluster)
+		switch {
+		case status.IsSuccess(): // Do nothing.
+		case status.IsInteralError():
+			return status
+		case status.IsClusterUnschedulable():
+			return status
+		default:
+			// Any status that is not Success, InternalError, or ClusterUnschedulable is considered an error.
+			return FromError(fmt.Errorf("filter plugin returned an unknown status %s", status), pl.Name())
+		}
+	}
+
+	return nil
+}
+
+// filteredClusterWithStatus is struct that documents clusters filtered out at the Filter stage,
+// along with a plugin status, which documents why a cluster is filtered out.
+//
+// This struct is used for the purpose of keeping reasons for returning scheduling decision to
+// the user.
+type filteredClusterWithStatus struct {
+	cluster *fleetv1beta1.MemberCluster
+	status  *Status
+}
+
+// runFilterPlugins runs filter plugins on clusters in parallel.
+func (f *framework) runFilterPlugins(ctx context.Context, state *CycleState, policy *fleetv1beta1.ClusterSchedulingPolicySnapshot, clusters []fleetv1beta1.MemberCluster) (passed []*fleetv1beta1.MemberCluster, filtered []*filteredClusterWithStatus, err error) {
+	// Create a child context.
+	childCtx, cancel := context.WithCancel(ctx)
+
+	// Pre-allocate slices to avoid races.
+	passed = make([]*fleetv1beta1.MemberCluster, len(clusters))
+	var passedIdx int32 = -1
+	filtered = make([]*filteredClusterWithStatus, len(clusters))
+	var filteredIdx int32 = -1
+
+	errFlag := parallelizer.NewErrorFlag()
+
+	doWork := func(pieces int) {
+		cluster := clusters[pieces]
+		status := f.runFilterPluginsFor(childCtx, state, policy, &cluster)
+		switch {
+		case status.IsSuccess():
+			// Use atomic add to avoid races with minimum overhead.
+			newPassedIdx := atomic.AddInt32(&passedIdx, 1)
+			passed[newPassedIdx] = &cluster
```

Comment:
should we use atomic.LoadInt32 to read the latest passedIdex?

Reply:
Not really. It would break the consistency.

Reply:
oh, the return value is already latest